### PR TITLE
TypeIdHasher: hash usize as u64

### LINKED ISSUE #36793: Regression in cross-compiling from 64 -> 32 bit
In today's most recent nightly:

```
rustc 1.13.0-nightly (d0623cf7b 2016-09-26)
binary: rustc
commit-hash: d0623cf7bda44849ab5df78a06b22f9108cf821a
commit-date: 2016-09-26
host: x86_64-unknown-linux-gnu
release: 1.13.0-nightly
```

Compiling a simple program:

```rust
fn main() {}
```

yields

```
$ rustc +nightly foo.rs --target i686-unknown-linux-gnu
error: linking with `cc` failed: exit code: 1
  |
  = note: "cc" "-Wl,--as-needed" "-Wl,-z,noexecstack" "-m32" "-L" "/home/alex/.multirust/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/i686-unknown-linux-gnu/lib" "foo.0.o" "-o" "foo" "-Wl,--gc-sections" "-pie" "-nodefaultlibs" "-L" "/home/alex/.multirust/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/i686-unknown-linux-gnu/lib" "-Wl,-Bstatic" "-Wl,-Bdynamic" "/home/alex/.multirust/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/i686-unknown-linux-gnu/lib/libstd-411f48d3.rlib" "/home/alex/.multirust/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/i686-unknown-linux-gnu/lib/libpanic_unwind-411f48d3.rlib" "/home/alex/.multirust/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/i686-unknown-linux-gnu/lib/libunwind-411f48d3.rlib" "/home/alex/.multirust/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/i686-unknown-linux-gnu/lib/librand-411f48d3.rlib" "/home/alex/.multirust/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/i686-unknown-linux-gnu/lib/libcollections-411f48d3.rlib" "/home/alex/.multirust/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/i686-unknown-linux-gnu/lib/librustc_unicode-411f48d3.rlib" "/home/alex/.multirust/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/i686-unknown-linux-gnu/lib/liballoc-411f48d3.rlib" "/home/alex/.multirust/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/i686-unknown-linux-gnu/lib/liballoc_jemalloc-411f48d3.rlib" "/home/alex/.multirust/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/i686-unknown-linux-gnu/lib/liblibc-411f48d3.rlib" "/home/alex/.multirust/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/i686-unknown-linux-gnu/lib/libcore-411f48d3.rlib" "/home/alex/.multirust/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/i686-unknown-linux-gnu/lib/libcompiler_builtins-411f48d3.rlib" "-l" "dl" "-l" "pthread" "-l" "gcc_s" "-l" "pthread" "-l" "c" "-l" "m" "-l" "rt" "-l" "util"
  = note: foo.0.o: In function `main':
foo.cgu-0.rs:(.text.main+0x2d): undefined reference to `std::rt::lang_start::h95854775e4ac8b95'
collect2: error: ld returned 1 exit status


error: aborting due to previous error
```

Seems odd to have an undefined symbol in the standard library! Taking a look at what's actually inside:

```
$ nm -g /home/alex/.multirust/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/i686-unknown-linux-gnu/lib/libstd-411f48d3.rlib | grep lang_start | c++filt
00000000 T std::rt::lang_start::hb5404024d2e88016
```

Looks like the hashes are different!

Did we perhaps leak a usize into metadata/symbol calculation?

cc @eddyb 
cc @michaelwoerister

### PR DESCRIPTION
**Pull request by TimNN**
Fixes #36793.

### CODE REVIEW

File: src/librustc/ty/util.rs [hunk 1]
```diff
@@ -436,7 +436,7 @@ impl<'a, 'gcx, 'tcx, H: Hasher> TypeVisitor<'tcx> for TypeIdHasher<'a, 'gcx, 'tc
             TyInt(i) => self.hash(i),
             TyUint(u) => self.hash(u),
             TyFloat(f) => self.hash(f),
-            TyArray(_, n) => self.hash(n),
+            TyArray(_, n) => self.hash(n as u64),
```
Comment (eddyb) [inline src/librustc/ty/util.rs]: Nice catch! Wait, why do we even do this?! `TyArray` should've held `u64` from the beginning.

**Comment by alexcrichton** [comment]
@bors: p=1

(fixing a regression)

Thanks @TimNN!

**Comment by eddyb** [comment]
Nominating for backport to 1.13 beta on the basis of https://github.com/rust-lang/rust/issues/36793#issuecomment-250443075 (needed by stage0 cross-compile).

**Comment by alexcrichton** [comment]
Regarding the nomination here, unfortunately it looks like this wasn't enough to fix https://github.com/rust-lang/rust/issues/36793 but the support in https://github.com/rust-lang/rust/pull/36866 I believe can be backported independently from this.

**Comment by nikomatsakis** [comment]
@alexcrichton are you saying this should not be backported, or just that it is not itself enough and we must backport #36866 too?

**Comment by alexcrichton** [comment]
@nikomatsakis I believe this should not be backported as https://github.com/rust-lang/rust/pull/36866 ended up undoing it after more cases of hashing `usize` were discovered.

**Comment by nikomatsakis** [comment]
@alexcrichton ok.

**Comment by nikomatsakis** [comment]
Based on @alexcrichton's comments, decided not to backport this PR.

cc @rust-lang/compiler

**Pull request merged**